File: app/services/audit_service.py
from __future__ import annotations

import uuid
from typing import List

import structlog
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import get_settings
from app.core.enums import AuditStatus
from app.core.exceptions import ForbiddenException, NotFoundException
from app.repositories.audit_repo import AuditRepository
from app.repositories.project_repo import ProjectRepository
from app.repositories.report_repo import ReportRepository
from app.schemas.audit import (
    AuditCreate,
    AuditListResponse,
    AuditResponse,
    AuditSummaryResponse,
    PageResponse,
    SeverityBreakdown,
    ViolationResponse,
)

logger = structlog.get_logger(__name__)
settings = get_settings()


class AuditService:
    def __init__(self, db: AsyncSession):
        self.db = db
        self.audit_repo = AuditRepository(db)
        self.project_repo = ProjectRepository(db)
        self.report_repo = ReportRepository(db)

    async def create_audit(self, user_id: str, payload: AuditCreate) -> tuple[AuditResponse, str]:
        """
        Creates an audit record and resolves the target URL.
        Returns (AuditResponse, resolved_target_url) so the router can
        fire the pipeline background task with the correct URL.
        """
        project = await self.project_repo.get_by_id(payload.project_id)
        if not project:
            raise NotFoundException("Project", str(payload.project_id))
        if str(project.user_id) != user_id:
            raise ForbiddenException()

        target_url = str(payload.target_url) if payload.target_url else project.base_url

        audit = await self.audit_repo.create(project_id=payload.project_id)
        logger.info(
            "Audit created",
            audit_id=str(audit.id),
            project_id=str(payload.project_id),
            target_url=target_url,
        )
        return AuditResponse.model_validate(audit), target_url

    async def list_audits(self, user_id: str) -> list[AuditListResponse]:
        """List all audits for the user's projects."""
        audits = await self.audit_repo.list_by_user(user_id)

        project_repo = ProjectRepository(self.db)

        result = []
        for audit in audits:
            project = await project_repo.get_by_id(audit.project_id)
            project_name = project.name if project else "Unknown Project"

            # Check for report to get accessibility score
            report = await self.report_repo.get_by_audit_id(audit.id)
            score = report.accessibility_score if report else None

            result.append(AuditListResponse(
                id=audit.id,
                project_id=audit.project_id,
                project_name=project_name,
                status=audit.status,
                error_message=audit.error_message,
                started_at=audit.started_at,
                completed_at=audit.completed_at,
                accessibility_score=score,
            ))

        return result

    async def get_audit(self, user_id: str, audit_id: uuid.UUID) -> AuditResponse:
        audit = await self._get_owned_audit(user_id, audit_id)
        return AuditResponse.model_validate(audit)

    async def get_summary(self, user_id: str, audit_id: uuid.UUID) -> AuditSummaryResponse:
        """
        Returns a combined audit + report summary.
        Report fields are None while the pipeline is still running.
        """
        audit = await self._get_owned_audit(user_id, audit_id)
        report = await self.report_repo.get_by_audit_id(audit_id)

        return AuditSummaryResponse(
            audit_id=audit.id,
            project_id=audit.project_id,
            status=audit.status,
            started_at=audit.started_at,
            completed_at=audit.completed_at,
            error_message=audit.error_message,
            accessibility_score=report.accessibility_score if report else None,
            pages_scanned=report.pages_scanned if report else None,
            total_violations=report.total_violations if report else None,
            severity_breakdown=SeverityBreakdown(
                critical=report.critical_count,
                serious=report.serious_count,
                moderate=report.moderate_count,
                minor=report.minor_count,
            ) if report else None,
            summary_text=report.summary_text if report else None,
        )

    async def get_pages(self, user_id: str, audit_id: uuid.UUID) -> List[PageResponse]:
        await self._get_owned_audit(user_id, audit_id)
        pages = await self.audit_repo.get_pages(audit_id)
        return [PageResponse.model_validate(p) for p in pages]

    async def get_violations(self, user_id: str, audit_id: uuid.UUID) -> List[ViolationResponse]:
        await self._get_owned_audit(user_id, audit_id)
        violations = await self.audit_repo.get_violations(audit_id)
        return [ViolationResponse.model_validate(v) for v in violations]

    # ---------------------------------------------------------------------------
    # Private
    # ---------------------------------------------------------------------------

    async def _get_owned_audit(self, user_id: str, audit_id: uuid.UUID):
        audit = await self.audit_repo.get_by_id(audit_id)
        if not audit:
            raise NotFoundException("Audit", str(audit_id))
        project = await self.project_repo.get_by_id(audit.project_id)
        if not project or str(project.user_id) != user_id:
            raise ForbiddenException()
        return audit

File: frontend/src/services/auditService.ts
import api from './api';

export interface StartAuditRequest {
  project_id: string;
  target_url?: string;
}

export interface AuditResponse {
  id: string;
  project_id: string;
  status: string;
  error_message: string | null;
  started_at: string;
  completed_at: string | null;
}

export interface AuditListResponse {
  id: string;
  project_id: string;
  project_name: string;
  status: string;
  error_message: string | null;
  started_at: string;
  completed_at: string | null;
  accessibility_score: number | null;
}

export interface AuditSummaryResponse {
  audit_id: string;
  project_id: string;
  status: string;
  started_at: string;
  completed_at: string | null;
  error_message: string | null;
  accessibility_score: number | null;
  pages_scanned: number | null;
  total_violations: number | null;
  severity_breakdown: {
    critical: number;
    serious: number;
    moderate: number;
    minor: number;
  } | null;
  summary_text: string | null;
}

export interface PageResponse {
  id: string;
  audit_id: string;
  url: string;
  title: string;
  violation_count: number;
  crawled_at: string;
}

export interface ViolationResponse {
  id: string;
  page_id: string;
  rule_id: string;
  severity: string;
  html_snippet: string | null;
  selector: string | null;
  wcag_criteria: string | null;
  ai_explanation: string | null;
  ai_fix: string | null;
  fix_type: string | null;
  ai_simulation: string | null;
  disability_types: string[];
  created_at: string;
}

export const auditService = {
  list: async (): Promise<AuditListResponse[]> => {
    const response = await api.get<AuditListResponse[]>('/audits');
    return response.data;
  },

  start: async (data: StartAuditRequest): Promise<AuditResponse> => {
    const response = await api.post<AuditResponse>('/audits', data);
    return response.data;
  },

  get: async (id: string): Promise<AuditResponse> => {
    const response = await api.get<AuditResponse>(`/audits/${id}`);
    return response.data;
  },

  getSummary: async (id: string): Promise<AuditSummaryResponse> => {
    const response = await api.get<AuditSummaryResponse>(`/audits/${id}/summary`);
    return response.data;
  },

  getPages: async (id: string): Promise<PageResponse[]> => {
    const response = await api.get<PageResponse[]>(`/audits/${id}/pages`);
    return response.data;
  },

  getViolations: async (id: string): Promise<ViolationResponse[]> => {
    const response = await api.get<ViolationResponse[]>(`/audits/${id}/violations`);
    return response.data;
  },
};